src/services/codefixes/inferFromUsage.ts

```diff
@@ -834,92 +905,158 @@ namespace ts.codefix {
+                    members.set(name, symbol);
+                });
             }
+            const callSignatures: Signature[] = usage.calls ? [getSignatureFromCalls(usage.calls)] : [];
+            const constructSignatures: Signature[] = usage.constructs ? [getSignatureFromCalls(usage.constructs)] : [];
+            const stringIndexInfo = usage.stringIndex && checker.createIndexInfo(combineFromUsage(usage.stringIndex), /*isReadonly*/ false);
+            return checker.createAnonymousType(/*symbol*/ undefined!, members, callSignatures, constructSignatures, stringIndexInfo, /*numberIndexInfo*/ undefined); // TODO: GH#18217
+        }
 
-            if (usage.numberIndex) {
-                types.push(checker.createArrayType(recur(usage.numberIndex)));
-            }
-            else if (usage.properties || usage.calls || usage.constructs || usage.stringIndex) {
-                const members = createUnderscoreEscapedMap<Symbol>();
-                const callSignatures: Signature[] = [];
-                const constructSignatures: Signature[] = [];
-                let stringIndexInfo: IndexInfo | undefined;
-
-                if (usage.properties) {
-                    usage.properties.forEach((u, name) => {
-                        const symbol = checker.createSymbol(SymbolFlags.Property, name);
-                        symbol.type = recur(u);
-                        members.set(name, symbol);
-                    });
-                }
+        function inferNamedTypesFromProperties(usage: Usage): Type[] {
+            if (!usage.properties || !usage.properties.size) return [];
+            const types = builtins.filter(t => allPropertiesAreAssignableToUsage(t, usage));
+            if (0 < types.length && types.length < 3) {
+                return types.map(t => inferInstantiationFromUsage(t, usage));
+            }
+            return [];
+        }
 
-                if (usage.calls) {
-                    for (const call of usage.calls) {
-                        callSignatures.push(getSignatureFromCall(call));
-                    }
+        function allPropertiesAreAssignableToUsage(type: Type, usage: Usage) {
+            if (!usage.properties) return false;
+            let result = true;
+            usage.properties.forEach((propUsage, name) => {
+                const source = checker.getTypeOfPropertyOfType(type, name as string);
+                if (!source) {
+                    result = false;
+                    return;
                 }
-
-                if (usage.constructs) {
-                    for (const construct of usage.constructs) {
-                        constructSignatures.push(getSignatureFromCall(construct));
-                    }
+                if (propUsage.calls) {
+                    const sigs = checker.getSignaturesOfType(source, SignatureKind.Call);
+                    result = result && !!sigs.length && checker.isTypeAssignableTo(source, getFunctionFromCalls(propUsage.calls));

```

Comment:
Here’s an interesting test case:

```ts
function test(a) {
  a.toString(2)
}
```

**Expected result**: `a` is `number`
**Actual result**: `a` is `{ toString: (arg0: number) => void; }`

It happens because every builtin’s `toString` call signature is assignable to the inferred structural one, and then the inference to builtins bails because of the limit of 3 union constituents. I briefly thought `source` and `target` might be backwards here—that you actually want to see if the _usage_ is assignable to the _builtin’s property_—but that’s too restrictive; the parameter type of `arg0: number` isn’t assignable to `radix?: number` under `strictFunctionTypes` because of the optionality of the latter, and the inferred return type `void` is not assignable to `string`. (Interestingly, if you swap `target` and `source`, turn off `strictFunctionTypes`, and change the line to `let x = a.toString(2)`, the inference to `number` occurs.)

I was able to hack up a fix for this something like:

```ts
const sigs = checker.getSignaturesOfType(source, SignatureKind.Call);
const usageType = getFunctionFromCalls(propUsage.calls);
const maxSourceParams = sigs.reduce((max, sig) => Math.max(max, sig.parameters.length), 0);
const maxUsageParams = usageType.getCallSignatures().reduce((max, sig) => Math.max(max, sig.parameters.length), 0);
result = result && !!sigs.length && maxUsageParams <= maxSourceParams && checker.isTypeAssignableTo(source, usageType);
```

but admittedly it feels kind of bad that checking assignability doesn’t Just Work™. Maybe I’m missing something clever?

Reply:
Errr, could we get better results by checking subtype first, then assignability, like we do for overload resolution?

Reply:
I think the root problem is that we expect to infer signatures whose lengths match, and have that beat the signatures whose lengths don't, right? I think an rule in inferFromUsage would be better than trying to get subtype or assignability to do the right thing, since they have other concerns to worry about -- they are part of the Real World Checker that has to be correct, but I feel like this preference is part of the Dream World Inference we have going on here, where you choose the best candidate on what feels right a lot of the time.

This would require maintaining some additional information -- specifically a list of source signatures matched with usage-signatures, and then that matched list would be used to rank the built-in types instead of just using `filter`. Basically, `combineNamedTypes`, by analogy with `combineTypes`. Maybe the decision could be even deferred to `combineTypes`.

Let's take this PR as-is and I'll think about it for the future.

Reply:
Oh, I mean, I just suggested subtype before assignability because i thought the difference in rules would already do the right thing (tm).